clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/op/manifest/CloudrunDeployManifestOperation.java

```diff
@@ -186,6 +214,101 @@ private List<String> insertSpinnakerAppNameServiceNameVersionName(
+                  if (yamlObj.getMetadata() != null) {
+                    LinkedHashMap<String, Object> metadataMap = yamlObj.getMetadata();
+                    LinkedHashMap<String, Object> specMap = yamlObj.getSpec();
+                    if (metadataMap != null && specMap != null) {
+                      if (metadataMap.get("annotations") != null
+                          && ((LinkedHashMap<String, Object>) specMap.get("template")).get("spec")
+                              != null) {
+                        LinkedHashMap<String, Object> specSpecMap =
+                            (LinkedHashMap<String, Object>)
+                                ((LinkedHashMap<String, Object>) specMap.get("template"))
+                                    .get("spec");
+                        LinkedHashMap<String, Object> annotationsMap =
+                            (LinkedHashMap<String, Object>) metadataMap.get("annotations");
+                        bindTheRequiredArtifact(annotationsMap, specSpecMap, artifacts);
+                      }
+                    }
+                  }
+                }
+                return yamlReader.writeValueAsString(yamlObj);
+              } catch (JsonProcessingException e) {
+                throw new RuntimeException(e);
+              }
+            })
+        .collect(Collectors.toList());
+  }
+
+  private void bindTheRequiredArtifact(
+      LinkedHashMap<String, Object> annotationsMap,
+      LinkedHashMap<String, Object> specSpecMap,
+      Collection<Artifact> artifacts) {
+
+    if (specSpecMap.get("containers") != null) {
+      List containerList = (List) specSpecMap.get("containers");
+      if (containerList != null && !containerList.isEmpty()) {
+        LinkedHashMap<String, Object> containerMap =
+            (LinkedHashMap<String, Object>) containerList.get(0);
+        for (Artifact artifact : artifacts) {
+          if (artifact.getType().equals("docker/image")) {
+            String cloudrunImage = (String) annotationsMap.get("client.knative.dev/user-image");
+            if (cloudrunImage != null) {
+              String[] imageArray = cloudrunImage.split(":");
+              String image = imageArray[0];
+              if (image != null && artifact.getName() != null) {
+                String[] imageArr = image.split("/");
+                String[] artifactArr = artifact.getName().split("/");
+                if (imageArr != null
+                    && artifactArr != null
+                    && imageArr.length > 0
+                    && artifactArr.length > 0) {
+                  String appImage = imageArr[imageArr.length - 1];
+                  String artifactImage = artifactArr[artifactArr.length - 1];
+                  if (appImage != null && artifactImage != null && appImage.equals(artifactImage)) {
+                    annotationsMap.put("client.knative.dev/user-image", artifact.getReference());
+                    containerMap.put("image", artifact.getReference());
```

Comment:
This if statement is incredibly difficult to read. Can we change the usage of `LinkedHashMap`s to well typed POJOs?